mrjob/emr.py

```diff
@@ -2627,33 +2627,44 @@ def _check_cluster_state(self, emr_client, cluster, num_steps):
 
         :return: -1 on failure or num_steps_in_cluster on success
         """
-        steps = list(_boto3_paginate(
-            'Steps',
-            emr_client, 'list_steps', ClusterId=cluster['Id']))
-
         if self._opts['release_label']:
             max_steps = _4_X_MAX_STEPS
+            allows_unlimited_steps = True
         else:
             image_version = cluster.get('RunningAmiVersion', '')
             max_steps = map_version(image_version, _IMAGE_VERSION_TO_MAX_STEPS)
+            allows_unlimited_steps = False
+
+        if self._opts['no_steps_limit'] and allows_unlimited_steps:
```

Comment:
Should we do the same thing in _get_job_steps??

Reply:
ListSteps returns jobs in reverse sorted order, so the submitted ones should be the first few and the logic aborts when something else is found. So it is already just as efficient.

Also, as ListSteps only supports querying by step states and not by name prefix, I don't see any other way of doing this.